File: spring-security-oauth2-demo/auth-server/src/main/java/com/sivalabs/authserver/service/SecurityUserDetailsService.java
package com.sivalabs.authserver.service;

import com.sivalabs.authserver.entity.UserAccount;
import com.sivalabs.authserver.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class SecurityUserDetailsService implements UserDetailsService {

    private final UserRepository userRepository;

    @Autowired
    public SecurityUserDetailsService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        Optional<UserAccount> optionalUserAccount = this.userRepository.findByEmail(username);
        if(!optionalUserAccount.isPresent()) {
            throw new UsernameNotFoundException("User not found");
        }
        return optionalUserAccount.get();
    }
}

File: spring-boot-k8s-demo/src/main/kotlin/com/sivalabs/geeksclub/security/SecurityUserDetailsService.kt
package com.sivalabs.geeksclub.security

import com.sivalabs.geeksclub.entities.User
import com.sivalabs.geeksclub.users.UserRepository
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
class SecurityUserDetailsService(val userRepository: UserRepository) : UserDetailsService {
    override fun loadUserByUsername(username: String): UserDetails {
        val user: Optional<User> = userRepository.findByEmail(username)
        if(!user.isPresent) {
            throw UsernameNotFoundException("User email $username not found")
        }
        return SecurityUser(user.get())
    }
}